From 7e67840943242b14e719072fefc65c6260634d22 Mon Sep 17 00:00:00 2001
From: Darren Kenny <darren.kenny@oracle.com>
Date: Thu, 5 Nov 2020 10:57:14 +0000
Subject: gnulib/regexec: Fix possible null-dereference

It appears to be possible that the mctx->state_log field may be NULL,
and the name of this function, clean_state_log_if_needed(), suggests
that it should be checking that it is valid to be cleaned before
assuming that it does.

Fixes: CID 86720

Signed-off-by: Darren Kenny <darren.kenny@oracle.com>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

Patch-Name: 2021-02-security/024-gnulib-regexec-Fix-possible-null-dereference.patch
---
 conf/Makefile.extra-dist                     |  1 +
 grub-core/gnulib-fix-regexec-null-deref.diff | 12 ++++++++++++
 grub-core/gnulib/regexec.c                   |  3 +++
 3 files changed, 16 insertions(+)
 create mode 100644 grub-core/gnulib-fix-regexec-null-deref.diff

diff --git a/conf/Makefile.extra-dist b/conf/Makefile.extra-dist
index 6ab5005b9..a22f7f016 100644
--- a/conf/Makefile.extra-dist
+++ b/conf/Makefile.extra-dist
@@ -30,6 +30,7 @@ EXTRA_DIST += grub-core/genemuinitheader.sh
 
 EXTRA_DIST += grub-core/gnulib-fix-null-deref.diff
 EXTRA_DIST += grub-core/gnulib-fix-null-state-deref.diff
+EXTRA_DIST += grub-core/gnulib-fix-regexec-null-deref.diff
 EXTRA_DIST += grub-core/gnulib-fix-uninit-structure.diff
 EXTRA_DIST += grub-core/gnulib-fix-unused-value.diff
 EXTRA_DIST += grub-core/gnulib-fix-width.diff
diff --git a/grub-core/gnulib-fix-regexec-null-deref.diff b/grub-core/gnulib-fix-regexec-null-deref.diff
new file mode 100644
index 000000000..50dddb25f
--- /dev/null
+++ b/grub-core/gnulib-fix-regexec-null-deref.diff
@@ -0,0 +1,12 @@
+--- a/lib/regexec.c	2020-10-21 14:25:35.310195912 +0000
++++ b/lib/regexec.c	2020-11-05 10:55:09.621542984 +0000
+@@ -1754,6 +1754,9 @@
+ {
+   Idx top = mctx->state_log_top;
+
++  if (mctx->state_log == NULL)
++    return REG_NOERROR;
++
+   if ((next_state_log_idx >= mctx->input.bufs_len
+        && mctx->input.bufs_len < mctx->input.len)
+       || (next_state_log_idx >= mctx->input.valid_len
diff --git a/grub-core/gnulib/regexec.c b/grub-core/gnulib/regexec.c
index 9264f2628..fdacff12c 100644
--- a/grub-core/gnulib/regexec.c
+++ b/grub-core/gnulib/regexec.c
@@ -1754,6 +1754,9 @@ clean_state_log_if_needed (re_match_context_t *mctx, Idx next_state_log_idx)
 {
   Idx top = mctx->state_log_top;
 
+  if (mctx->state_log == NULL)
+    return REG_NOERROR;
+
   if ((next_state_log_idx >= mctx->input.bufs_len
        && mctx->input.bufs_len < mctx->input.len)
       || (next_state_log_idx >= mctx->input.valid_len
